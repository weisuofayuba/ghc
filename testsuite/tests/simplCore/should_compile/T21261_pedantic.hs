{-# OPTIONS_GHC -fpedantic-bottoms #-} -- This flag must inhibit eta reduction based on demands

module T21261_pedantic where

-- README: See T21261. These examples absolutely should not eta reduce with
-- -fpedantic-bottoms.

f1 :: (Int -> Int -> Int) -> Int
f1 c = c 1 `seq` c 2 3
{-# NOINLINE f1 #-}
no2 :: (Int -> Int -> Int) -> Int
no2 c = f1 (\x y -> c x y)

f2 :: (Int -> Int -> Int) -> Maybe Int
f2 c = Just (c 1 `seq` c 3 4)
{-# NOINLINE f2 #-}
no2_lazy :: (Int -> Int -> Int) -> Maybe Int
no2_lazy c = f2 (\x y -> c x y)
