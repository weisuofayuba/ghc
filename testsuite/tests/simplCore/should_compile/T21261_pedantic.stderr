
==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 59, types: 63, coercions: 0, joins: 0/0}

lvl = I# 2#

lvl1 = I# 3#

lvl2 = I# 1#

f1 = \ c -> case c lvl2 of { __DEFAULT -> c lvl lvl1 }

no2 = \ c -> f1 (\ x y -> c x y)

lvl3 = I# 4#

$wf2 = \ c -> (# case c lvl2 of { __DEFAULT -> c lvl1 lvl3 } #)

f2 = \ c -> case $wf2 c of { (# ww #) -> Just ww }

no2_lazy
  = \ c -> case $wf2 (\ x y -> c x y) of { (# ww #) -> Just ww }



